[DeviceMesh] Simplifying internal bookkeeping with CuTe layout

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164510
* __->__ #163213

We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the `_MeshLayout` to handle all index operations rather use a map to record mesh dims.
2. Removed `flatten_name_to_root_dims`, because now we can directly get layout from a flattened device mesh.
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use the newly added function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe. The reason is that layout acts as a backend of mesh tensor bookkeeping (indexing indices), it needs to be used as indices for remap back to the mesh tensor for new DeviceMesh generation and backend init. For example, in the case of 2K to 4K, the underlying layout is (2K, 1) but the actual value of the mesh tensor is [2K, 2K+1, ....,]. While flattening, slicing, we need to remap the layout back to the new mesh tensor so it maps the actual device allocation. For example, in the 2K to 4K case, if the shape is (1K, 1K) with dim_names ("dp", "tp"). Then when slicing "tp", the mesh tensor should be (2K, 2K+1, ..., 3K-1) or (3K, 3K+1, ... 4K-1). not the global ranks generated from the layout. (1K, 1).

Verified that loss curve is very close for DeepSeekV3 on torchtitan, note that exact same match is challenging because even if we run the baseline twice, the loss curve does not exactly match.

<img width="1113" height="490" alt="image" src="https://github.com/user-attachments/assets/7877b5a4-337e-4ad8-b878-2378f4f0f38d" />


The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

fduwjj committed:
[DeviceMesh] Simplifying internal bookkeeping with CuTe layout

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (originally messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (originally messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -308,29 +274,27 @@ def _get_slice_mesh_dims(
-            return slice_mesh_dims
+            return layout_sliced
 
         def _get_all_submeshes(
```
it is used internally by shampoo. cc: @wz337

fegin replied:
Let's add TODO to these private APIs used by other components. This mostly represents our DeviceMesh should public these use cases in the future.

lw replied:
Could we get clarity on which context this is needed in and whether the same can be achieved in other ways?

ezyang replied:
Here's what the code usage in Shampoo looks like:

```
        # NOTE: We get all submeshes along the "replicate" dimension, then pick out
        # the sub-mesh that the optimizer state is assigned to.
        #
        # For the example above, this would give me submeshes [[3, 27], [11, 35], [19, 43]].
        # Note that the group source rank must belong to {0, 1, 2} in this case.
        # Suppose the group_source_rank = 1, then this would get the submesh [11, 35].
        replicate_submesh = _mesh_resources._get_all_submeshes(
            device_mesh_2d, "replicate"
        )[group_source_rank]
```

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -339,10 +303,10 @@ def _get_all_submeshes(
+            )  # type: ignore[arg-type]
             cur_rank = device_mesh.get_rank()
             res_submeshes = []
             for mesh_1d in pg_ranks_by_dim:
```
For this case, we are getting submesh per dim, so each dim is a 1d mesh? Even with nested one, we have to flatten it because per definition of mesh dim, that dimension only has one mesh.

fduwjj commented on torch/distributed/_mesh_layout.py:
```diff
@@ -210,3 +221,141 @@ def global_ranks(self, world_size: int) -> list[list[int]]:
+            previous_span = size * stride
+        return True
+
+    def to_remapping_tensor(
```
`remapping_tensor` is a name we agreed on during the discussion last Monday. but I am open to other proposals. Maybe can you elaborate more on what exact contexts you are missing here?

Basically what we wrote in the comment:
> With this method, the cute layout serves as the backend of indices bookkeeping for the
        mesh tensor when it comes to flatten, unflatten and slicing operations. The actual mesh
        tensor still represents the actual device assignment and ranks.
        
which part sound not clear to you? Or are asking why we want to do this conversion? It is because we need an actual mesh tensor to specify device allocation and also for backend creation.

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -352,11 +316,14 @@ def _get_all_submeshes(
                     if cur_rank in mesh_1d
                     else []
                 )
+                submesh._layout = (
+                    device_mesh._layout if cur_rank in mesh_1d else _MeshLayout(0, 0)
+                )
```
Actually after a second thought, we can remove this part, since a mesh maps to a layout, so this line of change is not needed.

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fegin commented on torch/distributed/device_mesh.py:
```diff
@@ -317,29 +271,27 @@ def _get_slice_mesh_dims(
+            layout_sliced = _MeshLayout(sliced_sizes, sliced_strides)
+            if not layout_sliced.check_non_overlap():
+                raise RuntimeError(
+                    f"slicing overlapping dim_names {mesh_dim_names} is not allowed"
```
```suggestion
                    f"Slicing overlapping dim_names {mesh_dim_names} is not allowed."
```

fduwjj commented on torch/distributed/_mesh_layout.py:
```diff
@@ -210,3 +221,144 @@ def global_ranks(self, world_size: int) -> list[list[int]]:
+        ):
+            return tensor
+
+        # This is important because the indices generated by the layout will be larger than the original mesh tensor
+        # when the original mesh tensor does not contain all ranks in the world. So we need to scale the layout's stride
+        # by world_size // mesh_tensor.numel() so that the indices generated by the layout will be within the range of
+        # the original mesh tensor.
+        if original_mesh_tensor.numel() != world_size:
```
yes, because we still needs to maintain the slicing of flattened tensor from root. That way we will need to keep the original layout not the layout derived from the submesh's mesh tensor.

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -65,84 +67,41 @@ def get_current_mesh(self) -> "DeviceMesh":
+                device_mesh.mesh,
+                get_world_size(),
+            )
             for mesh_nd in pg_ranks_by_dim:
                 submesh = DeviceMesh(
                     device_mesh.device_type,
                     mesh_nd,
                     mesh_dim_names=submesh_dim_names,
                     _init_backend=False,
+                    _layout=layout,
                 )
                 if cur_rank in mesh_nd:
                     res_submesh = submesh
```
Can you explain this code? I know that it wasn't added by this PR, but I find it odd. Basically, we create a potentially very large number of DeviceMesh objects (e.g., on 1024 GPUs with TP=2, this could be up to 512 DeviceMeshes), only to throw them all away except for one. Why do we do that?

I suspect this means that constructing a DeviceMesh objects must have some global side-effects. Is that so? What kind of effects are we talking about?

lw replied:
Note that in this snipped `init_backend` is False so we're not even creating sub-PGs...

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -180,50 +134,48 @@ def create_flatten_mesh(
-            ).reshape(-1, flattened_mesh_dim_size)
-
             cur_rank = root_mesh.get_rank()
+            # Due to the limitation of ProcessGroup api, we need to start from root mesh so that all ranks call the
+            # new_group api to avoid potential hang.
```
I feel bad that we need to make the DeviceMesh code more complicated in order to deal with these low-level concerns. Moreover, if the backend supports splitting (e.g., NCCL) then all of this isn't even actually needed!

May I suggest that we add a helper function that "splits" the root PG into an arbitrary "tiling" (i.e., N subgroups which cover everything) which abstracts away these complexities by internally deciding whether to call split or do this for loop?

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -52,9 +54,9 @@ def __init__(self) -> None:
             self.root_to_flatten_mapping: dict[DeviceMesh, dict[str, DeviceMesh]] = {}
-            # Record flatten mesh name to its mesh dim index in root mesh.
-            self.flatten_name_to_root_dims: dict[
-                DeviceMesh, dict[str, tuple[int, ...]]
+            # Record flatten mesh name to its flattened layout in root mesh.
+            self.flatten_name_to_root_layout: dict[
```
Can I ask if you intend to keep these global (thread-local) mappings, or if the plan is to eventually drop them? I still don't understand exactly why we need them, but if they're slated for removal I won't worry about them too much.

ezyang replied:
If this is legacy, help me understand why storing MeshLayout over mesh dim index is preferable? For legacy things I would have naively expected you to keep them exactly as is and route around the damage.

fegin replied:
Probably not in this PR, and I'm not sure if you have done this in the unflatten PR, but we should add some TODOs and description about why these code exist and not the same as the unflatten design. BC is mostly the reason.

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -317,41 +271,41 @@ def _get_slice_mesh_dims(
                     )
-                curr_idx = next_idx
+                elif name in flatten_name_to_root_layout:
+                    layout_sliced.append(flatten_name_to_root_layout[name])
```
Should there be an `else` here to catch remaining cases? (Maybe just throwing)

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -317,41 +271,41 @@ def _get_slice_mesh_dims(
+                elif name in flatten_name_to_root_layout:
+                    layout_sliced.append(flatten_name_to_root_layout[name])
+
+            sliced_sizes = tuple(l.sizes for l in layout_sliced)
+            sliced_strides = tuple(l.strides for l in layout_sliced)
+            # When users sliced dim_names outside from current mesh, we will check whether
+            # there is layout overlap.
+            # TODO: Eventually we will just directly throw error here because
+            # we will deprecate the slicing of flattened dim_name from root mesh.
+            layout_sliced = _MeshLayout(sliced_sizes, sliced_strides)
+            if not layout_sliced.check_non_overlap():
+                raise RuntimeError(
+                    f"Slicing overlapping dim_names {mesh_dim_names} is not allowed."
+                )
```
Nice, this has become very clean and simple!

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -317,41 +271,41 @@ def _get_slice_mesh_dims(
+            layout_sliced = []
+            for name in mesh_dim_names:
+                if name in device_mesh.mesh_dim_names:
+                    layout_sliced.append(
+                        device_mesh._layout[device_mesh.mesh_dim_names.index(name)]
                     )
-                curr_idx = next_idx
+                elif name in flatten_name_to_root_layout:
```
Maybe this will get addressed in later PRs, but what I see here is that we make a distinction between two "types" of dimensions: the ones that come from flattening and all the other ones. This distinction feels artificial to me. Once a dimension has been flattened, we compute its (nested?) sizes and strides, just like any other dimension, and then it should become indistinguishable and we should have a single code path. Why does it still require special handling?

lw replied:
I agree with @ezyang, it'd be very valuable to get a clear description of the legacy constraints (documented in the code, not as GitHub comments) so we can keep them in mind when proposing changes.

As I understand it, the legacy behavior you're describing occurs during lookup of a named mesh dim (i.e., inside the `__getitem__` method): on the _root_ mesh (and only on that one!) needs to be aware of dimensions that are not part of the original dims at creation, but which were obtained via flattening some of them. Is that so?

If that's the case then I'd expect to see some ad-hoc code inside the `__getitem__` method (plus some bookkeeping elsewhere), but the rest of the code to be relatively unimpacted.

fduwjj replied:
Yes it is because we need `__getitem__` to work with flatten for BC reason. And to clarify, this requires the bookeeping to have hack both at flatten and `__getitem__`.

lw reviewed (approved):
This is going in the right direction, LGTM! I left some comments on how I think we can push this further, but perhaps you're already taking care of them in the rest of the stack

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Separated the Devicemesh creation via mesh into a util function `_create_mesh_from_ranks`.
7. Added a type alias for `tuple[Optional[str], Optional[C10dBackend.Options]]` to be `BackendConfig`


The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Separated the Devicemesh creation via mesh into a util function `_create_mesh_from_ranks`.
7. Added a type alias for `tuple[Optional[str], Optional[C10dBackend.Options]]` to be `BackendConfig`


The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -65,17 +65,18 @@ def _init_device_mesh_stub():
-            self.mesh_dim_group_options: dict[
-                int, tuple[Optional[str], Optional[C10dBackend.Options]]
-            ] = {}
+            self.mesh_dim_group_options: dict[int, BackendConfig] = {}
```
Just for reference, this is something you could put in an earlier stacked PR, easy BE and easy stamp

fduwjj replied:
Split and merged in https://github.com/pytorch/pytorch/pull/163928

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -477,6 +428,20 @@ def __init__(
+            self._layout = (
+                _layout
+                if _layout
+                else _MeshLayout(self.mesh.size(), self.mesh.stride())
```
Does the layout always have to be consistent with self.mesh.size()/stride()?

fduwjj replied:
Its numel has to be the same, but the value of size and stride no. That's exactly why we need to scale in the remapping.

fduwjj replied:
So in an ideal world, layout always has to be consistent with self.mesh.size()/stride(). But currently we have couple constraints:
1. All ranks need to call backend init, even if that ranks are not part it. So we will want to keep the layout to be same as before the submesh is sliced. This is also why we need to do scale.
2. Another thing we need to support slicing out flattened mesh from root mesh.

We believe both will be gone eventually and we will be able to eventually come back to this consistency and get rid of scaling as well.

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Separated the Devicemesh creation via mesh into a util function `_create_mesh_from_ranks`.
7. Added a type alias for `tuple[Optional[str], Optional[C10dBackend.Options]]` to be `BackendConfig`


The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Separated the Devicemesh creation via mesh into a util function `_create_mesh_from_ranks`.
7. Added a type alias for `tuple[Optional[str], Optional[C10dBackend.Options]]` to be `BackendConfig`


Another function (the name is open for discussion) is that layout acts as a backend of mesh tensor bookkeeping (indexing indices), it needs to be used as indices for remap back to the mesh tensor for new DeviceMesh generation and backend init. For example, in the case of 2K to 4K, the underlying layout is (2K, 1) but the actual value of the mesh tensor is [2K, 2K+1, ....,]. While flattening, slicing, we need to remap the layout back to the new mesh tensor so it maps the actual device allocation. For example, in the 2K to 4K case, if the shape is (1K, 1K) with dim_names ("dp", "tp"). Then when slicing "tp", the mesh tensor should be (2K, 2K+1, ..., 3K-1) or (3K, 3K+1, ... 4K-1). not the global ranks generated from the layout. (1K, 1).

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fegin commented on torch/distributed/_mesh_layout.py:
```diff
@@ -243,3 +244,90 @@ def check_non_overlap(self) -> bool:
+    def remap_to_tensor(
+        self,
+        mesh_tensor: torch.Tensor,
+        world_size: int,
```
If 1) we don't need all ranks to participate in sub pg creation, 2) all operation, including flatten and slicing are functional, can I claim that world_size should be needed for this function because both mesh_tensor and layout will be scaled appropriately. So world_size is always the same as mesh_tensor.numel() and we don't need to worry about sub PG creation.

fegin replied:
If my new understanding is correct, then my original comment was incorrect that this `remap_to_tensor()` can be a standalone PR because its argument is tightly coupled with DeviceMesh and PG details. Then we may need to describe the restriction (sub PG creation and slicing is not functional)

fegin commented on torch/distributed/device_mesh.py:
```diff
@@ -342,41 +288,41 @@ def _get_slice_mesh_dims(
-            return slice_mesh_dims
+            return layout_sliced
 
+        # TODO: to make this use case by other components public API in the future.
```
If we try to make all operations functional and only keep the `root` mesh tensor, is this still doable?

ezyang commented:
I still don't really understand the BC concerns but I am happy to defer to @lw and @fegin here!

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -66,16 +66,17 @@ def _init_device_mesh_stub():
             self.child_to_root_mapping: dict[DeviceMesh, DeviceMesh] = {}
             self.mesh_dim_group_options: dict[int, BackendConfig] = {}
             self.root_to_flatten_mapping: dict[DeviceMesh, dict[str, DeviceMesh]] = {}
-            # Record flatten mesh name to its mesh dim index in root mesh.
-            self.flatten_name_to_root_dims: dict[
-                DeviceMesh, dict[str, tuple[int, ...]]
+            # Record flatten mesh name to its flattened layout in root mesh.
+            self.flatten_name_to_root_layout: dict[
+                DeviceMesh, dict[str, _MeshLayout]
             ] = {}
```
[MAJOR] I really want to get rid of these attributes, as I dislike any form of global state. `mesh_dim_group_options` is already unofficially deprecated since https://github.com/pytorch/pytorch/pull/159371, so let's focus on the others. I believe that reworking this will help simplify the rest of the code too, so please hear me out.

The keys of both `root_to_flatten_mapping` and `flatten_name_to_root_layout` are all and only the _root_ meshes. So, instead of storing this in a single centralized location (as a dict), we could store this information directly as raw _attributes_ on the root mesh _objects_. Concretely:
```diff
- _mesh_resources.root_to_flatten_mapping[_mesh_resources.get_root_mesh(m)][mesh_dim]
+ _mesh_resources.get_root_mesh(m)._root_to_flatten_mapping[mesh_dim]
```

Now, we could add these attributes to the DeviceMesh class, which means that _all_ meshes get them, and we just leave these empty on non-root ones. However, I think there's a cleaner solution: we could define a new class, let's call it `SharedState` for now, and store it on each DeviceMesh under the `_shared_state` attribute. When we create a new _root_ mesh we initialize a new fresh ShareState for it, but when we create a _child_ mesh (via slicing, flattening, ...) we make it reuse the same SharedState as the parent. This way all meshes will have the same SharedState as their root. This allows to simplify the above expression even further:
```diff
- _mesh_resources.get_root_mesh(m)._root_to_flatten_mapping[mesh_dim]
+ m._shared_state._root_to_flatten_mapping[mesh_dim]
```
Note how we've managed to get rid of all global state! Yay!

With all this preliminary work, we can now leverage this SharedState class to simplify many more things:
- we could add a `root_mesh` attribute to SharedState, so that we replace the global state `_mesh_resources.child_to_root_mapping[m]` with `m._shared_state.root_mesh`.
- instead of storing `_dim_group_names` on _each_ mesh, and copying most of it over when we slice/flatten/..., we could have a common dictionary which maps the (sub-)MeshLayout to the name of the ProcessGroup that backs it;
- we could add a field that tracks which dimension names are the result of flattening, so that the `__getitem__` method on the root device mesh can access this and easily implement the legacy behavior;
- I think we could even centralize some of the name-to-(sub-)layout mapping, but let's maybe do the other changes first to see what they give.

Does this make sense? Do you see any issues with this proposal that I might have missed?

fduwjj replied:
I think the change you proposed here itself should be a separate PR because we will need to add one more comparison attribute to DeviceMesh Hash which could make DTensor CPU overhead increase. I mean we need to do benchmarking for this as well. Or you really feel strong about we should do this in this PR?

lw reviewed (commented):
This is a pretty big change so I want to make sure I get it. There are a few aspects I still don't understand, and I'd like to tackle them one by one to avoid getting confused. The first one is the infamous legacy behavior that we need to preserve, and how that is preventing us from simplifying the bookkeeping of flattened dims and remove global state. I left one major comment where I propose a step towards a simplification.

Once that's addressed, I'll also want to discuss the `remap_to_tensor` method, as I still don't get it fully.

ezyang reviewed (dismissed):
blocking merge pending BC considerations discussion

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -86,109 +87,49 @@ def get_current_mesh(self) -> "DeviceMesh":
-            res_submesh._dim_group_names = slice_dim_group_name  # type: ignore[possibly-undefined, has-type]
-            self.child_to_root_mapping[res_submesh] = device_mesh
-
+            res_submesh._dim_group_names = slice_dim_group_name
```
it is always List[str] because it is pg_group_name which is a string, it can be a str of a global counter (again c10d related) or it can be a hash of all ranks within a pg. This is controlled by a kwarg in generating pg_name. Code pointer: https://github.com/pytorch/pytorch/blob/95be302889b8683b7ec7793a69ffa8891b6b5af8/torch/distributed/distributed_c10d.py#L5017

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -208,47 +149,50 @@ def create_flatten_mesh(
-            ).reshape(-1, flattened_mesh_dim_size)
-
             cur_rank = root_mesh.get_rank()
+            # Due to the limitation of ProcessGroup api, we need to start from root mesh so that all ranks call the
+            # new_group api to avoid potential hang.
```
yes it is always like that unless we fix it on the c10d level

ezyang commented:
bouncing request changes per VC

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Add an util function (the name is open for discussion) is that layout acts as a backend of mesh tensor bookkeeping (indexing indices), it needs to be used as indices for remap back to the mesh tensor for new DeviceMesh generation and backend init. For example, in the case of 2K to 4K, the underlying layout is (2K, 1) but the actual value of the mesh tensor is [2K, 2K+1, ....,]. While flattening, slicing, we need to remap the layout back to the new mesh tensor so it maps the actual device allocation. For example, in the 2K to 4K case, if the shape is (1K, 1K) with dim_names ("dp", "tp"). Then when slicing "tp", the mesh tensor should be (2K, 2K+1, ..., 3K-1) or (3K, 3K+1, ... 4K-1). not the global ranks generated from the layout. (1K, 1).

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

lw reviewed (commented):
Thanks for the changes! I believe this has helped a lot the readability of the code!

I took the time to read this carefully in full. In doing so I believe I found 3-4 bugs which, apparently, weren't caught by the current test suite. We should add tests for them immediately, and fix them.

I also think that we're constructing too many unnecessary mesh tensors. This is an expensive operation (proportional to the world size) and the point of switching to CuTe layouts was precisely to avoid doing so. It should be extremely easy to remove these intermediate mesh tensors, provided we introduce a new (private) constructor for DeviceMesh.

fduwjj commented on torch/distributed/_mesh_layout.py:
```diff
@@ -243,3 +244,53 @@ def check_non_overlap(self) -> bool:
+        ranks_indices_list = self.global_ranks(mesh_tensor.numel())
+        indices = torch.tensor(ranks_indices_list, device="cpu", dtype=torch.int).view(
+            -1,
+            *flatten(self.sizes),  # type: ignore[arg-type]
```
hmmm let me think more on this one, we have mesh["dp_tp"] but we haven't tested the slicing with cp case.

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -208,47 +150,46 @@ def create_flatten_mesh(
-            self.child_to_root_mapping[res_flattened_mesh] = root_mesh  # type: ignore[possibly-undefined]
+            self.child_to_root_mapping[res_flattened_mesh] = root_mesh
             self.root_to_flatten_mapping.setdefault(root_mesh, {})[mesh_dim_name] = (
-                res_flattened_mesh  # type: ignore[possibly-undefined]
+                res_flattened_mesh
+            )
+            self.flatten_name_to_root_layout[root_mesh][mesh_dim_name] = (
+                res_flattened_mesh._layout
             )
```
that is true... I think previously with indexes, it is not possible to consolidate, but now with layout, we can do that. Another point to support having CuTe layout to simply internal bookkeeping.

fduwjj commented on test/distributed/test_device_mesh.py:
as discussed above, original device_mesh does not support this way of slicing, I will address this behavior in follow-up PR because this is a refactoring PR.

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Add an util function (the name is open for discussion) is that layout acts as a backend of mesh tensor bookkeeping (indexing indices), it needs to be used as indices for remap back to the mesh tensor for new DeviceMesh generation and backend init. For example, in the case of 2K to 4K, the underlying layout is (2K, 1) but the actual value of the mesh tensor is [2K, 2K+1, ....,]. While flattening, slicing, we need to remap the layout back to the new mesh tensor so it maps the actual device allocation. For example, in the 2K to 4K case, if the shape is (1K, 1K) with dim_names ("dp", "tp"). Then when slicing "tp", the mesh tensor should be (2K, 2K+1, ..., 3K-1) or (3K, 3K+1, ... 4K-1). not the global ranks generated from the layout. (1K, 1).

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe.
6. Add an util function (the name is open for discussion) is that layout acts as a backend of mesh tensor bookkeeping (indexing indices), it needs to be used as indices for remap back to the mesh tensor for new DeviceMesh generation and backend init. For example, in the case of 2K to 4K, the underlying layout is (2K, 1) but the actual value of the mesh tensor is [2K, 2K+1, ....,]. While flattening, slicing, we need to remap the layout back to the new mesh tensor so it maps the actual device allocation. For example, in the 2K to 4K case, if the shape is (1K, 1K) with dim_names ("dp", "tp"). Then when slicing "tp", the mesh tensor should be (2K, 2K+1, ..., 3K-1) or (3K, 3K+1, ... 4K-1). not the global ranks generated from the layout. (1K, 1).

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

yangw-dev commented:
@pytorchbot revert -c ghfirst -m "caused internal test failure"

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the `_MeshLayout` to handle all index operations rather use a map to record mesh dims.
2. Removed `flatten_name_to_root_dims`, because now we can directly get layout from a flattened device mesh.
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use the newly added function `check_overlap` to check layout overlap.
5. Use a new function `to_remapping_tensor` to use layout ranks as indices when the mesh tensor is not representable as CuTe. The reason is that layout acts as a backend of mesh tensor bookkeeping (indexing indices), it needs to be used as indices for remap back to the mesh tensor for new DeviceMesh generation and backend init. For example, in the case of 2K to 4K, the underlying layout is (2K, 1) but the actual value of the mesh tensor is [2K, 2K+1, ....,]. While flattening, slicing, we need to remap the layout back to the new mesh tensor so it maps the actual device allocation. For example, in the 2K to 4K case, if the shape is (1K, 1K) with dim_names ("dp", "tp"). Then when slicing "tp", the mesh tensor should be (2K, 2K+1, ..., 3K-1) or (3K, 3K+1, ... 4K-1). not the global ranks generated from the layout. (1K, 1).

Verified that loss curve is very close for DeepSeekV3 on torchtitan, note that exact same match is challenging because even if we run the baseline twice, the loss curve does not exactly match.

<img width="1113" height="490" alt="image" src="https://github.com/user-attachments/assets/7877b5a4-337e-4ad8-b878-2378f4f0f38d" />


The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

With this refactoring we also enabled the slicing and flatten of non-contiguous dims of a device mesh which is hard to implement without cute layout.

This is a continue of https://github.com/pytorch/pytorch/pull/161106 (original one got messed with EasyCLA)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
